source/common/http/http2/protocol_constraints.cc

```diff
@@ -0,0 +1,117 @@
+#include "common/http/http2/protocol_constraints.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http2 {
+
+ProtocolConstraints::ProtocolConstraints(
+    CodecStats& stats, const envoy::config::core::v3::Http2ProtocolOptions& http2_options)
+    : stats_(stats), max_outbound_frames_(http2_options.max_outbound_frames().value()),
+      frame_buffer_releasor_([this]() { releaseOutboundFrame(); }),
+      max_outbound_control_frames_(http2_options.max_outbound_control_frames().value()),
+      control_frame_buffer_releasor_([this]() { releaseOutboundControlFrame(); }),
+      max_consecutive_inbound_frames_with_empty_payload_(
+          http2_options.max_consecutive_inbound_frames_with_empty_payload().value()),
+      max_inbound_priority_frames_per_stream_(
+          http2_options.max_inbound_priority_frames_per_stream().value()),
+      max_inbound_window_update_frames_per_data_frame_sent_(
+          http2_options.max_inbound_window_update_frames_per_data_frame_sent().value()) {}
+
+ProtocolConstraints::ReleasorProc
+ProtocolConstraints::incrementOutboundFrameCount(bool is_outbound_flood_monitored_control_frame) {
+  ++outbound_frames_;
+  if (is_outbound_flood_monitored_control_frame) {
+    ++outbound_control_frames_;
+  }
+  if (status_.ok()) {
+    status_ = checkOutboundQueueLimits();
+  }
+  return is_outbound_flood_monitored_control_frame ? control_frame_buffer_releasor_
+                                                   : frame_buffer_releasor_;
+}
+
+void ProtocolConstraints::releaseOutboundFrame() {
+  ASSERT(outbound_frames_ >= 1);
+  --outbound_frames_;
+}
+
+void ProtocolConstraints::releaseOutboundControlFrame() {
+  ASSERT(outbound_control_frames_ >= 1);
+  --outbound_control_frames_;
+  releaseOutboundFrame();
+}
+
+Status ProtocolConstraints::checkOutboundQueueLimits() {
+  if (outbound_frames_ > max_outbound_frames_) {
+    stats_.outbound_flood_.inc();
```

Comment:
I think you're over counting these stats since you end up calling checkOutboundQueueLimits() from the increment method in cases where you're outputting to upstream instead of downstream.

Possible solution: remove status_ and call checkOutboundQueueLimits() from the downstream connection's checkOutboundQueueLimits() method.

Reply:
The check* methods are not called if the status has flipped. I have moved status check into the methods to make more readable. Also tests verify stats counts to make sure there is no overcounting.

Reply:
The comment is related processing in the upstream direction.  What prevents outbound flooding in the upstream direction from triggering and incrementing this stat?  I think we agree that it is possible to call ProtocolConstraints::checkOutboundQueueLimits in the upstream direction, and the overflow condition can happen in that case.

Reply:
Yes, it was a problem for outbound frame checks. Should be fixed now.